File: frontend/src/api/roster.ts
import apiClient from './client'

// --- Types ---

export interface RosterStats {
  all: number
  cadets: number
  members: number
  cadets_need_presentation: number
  cadets_ready_to_promote: number
}

export interface RosterUser {
  id: number
  nickname: string
  status: number
  status_as_string: string | null
  active_module_count: number
  need_presentation: boolean
  is_ready_to_promote: boolean
}

export interface RosterModule {
  id: number
  name: string
  long_name: string
  code: string
  type: number
  period: number | null
  period_as_string: string | null
  image_header_uuid: string | null
  user_count: number
}

export interface RosterModuleDetailUser {
  id: number
  nickname: string
  status: number
  status_as_string: string | null
  active: boolean
  level: number
  level_as_string: string | null
}

export interface RosterModuleDetail {
  module: RosterModule
  users: RosterModuleDetailUser[]
}

export interface OfficeMember {
  nickname: string
  status: number
  status_as_string: string | null
}

export interface OfficeData {
  president: OfficeMember | null
  president_deputy: OfficeMember | null
  treasurer: OfficeMember | null
  treasurer_deputy: OfficeMember | null
  secretary: OfficeMember | null
  secretary_deputy: OfficeMember | null
}

// --- API functions ---

export async function getOffice(): Promise<OfficeData> {
  const { data } = await apiClient.get<OfficeData>('/roster/office')
  return data
}

export async function getRosterStats(): Promise<RosterStats> {
  const { data } = await apiClient.get<RosterStats>('/roster/stats')
  return data
}

export async function getRosterPilots(group: string = 'all'): Promise<RosterUser[]> {
  const { data } = await apiClient.get<RosterUser[]>('/roster/pilots', { params: { group } })
  return data
}

export async function getRosterModules(type: number, group: string = 'all'): Promise<RosterModule[]> {
  const { data } = await apiClient.get<RosterModule[]>('/roster/modules', { params: { type, group } })
  return data
}

export async function getRosterModuleDetail(
  moduleId: number,
  group: string = 'all',
): Promise<RosterModuleDetail> {
  const { data } = await apiClient.get<RosterModuleDetail>(`/roster/modules/${moduleId}`, {
    params: { group },
  })
  return data
}

File: backend/app/schemas/roster.py
from pydantic import BaseModel, Field


class RosterStatsOut(BaseModel):
    all: int
    cadets: int
    members: int
    cadets_need_presentation: int = 0
    cadets_ready_to_promote: int = 0


class RosterUserOut(BaseModel):
    id: int
    nickname: str
    status: int
    status_as_string: str | None = None
    active_module_count: int = 0
    need_presentation: bool = False
    is_ready_to_promote: bool = False


class RosterModuleOut(BaseModel):
    id: int
    name: str
    long_name: str
    code: str
    type: int
    period: int | None = None
    period_as_string: str | None = None
    image_header_uuid: str | None = None
    user_count: int = 0
    total_group_count: int = 0


class RosterModuleDetailUserOut(BaseModel):
    id: int
    nickname: str
    status: int
    status_as_string: str | None = None
    active: bool
    level: int
    level_as_string: str | None = None


class RosterModuleDetailOut(BaseModel):
    module: RosterModuleOut
    users: list[RosterModuleDetailUserOut] = Field(default_factory=list)


class OfficeMemberOut(BaseModel):
    nickname: str
    status: int
    status_as_string: str | None = None


class OfficeOut(BaseModel):
    president: OfficeMemberOut | None = None
    president_deputy: OfficeMemberOut | None = None
    treasurer: OfficeMemberOut | None = None
    treasurer_deputy: OfficeMemberOut | None = None
    secretary: OfficeMemberOut | None = None
    secretary_deputy: OfficeMemberOut | None = None
